README
WebService-EastCoJp-Dictionary version 0.02
===========================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  - Modern::Perl
  - Moose
  - Readonly
  - URI::Escape
  - XML::XPath
  - Test::LongString

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension WebService::EastCoJp::Dictionary.

0.02  Sun Aug 29 19:31:13 2010
    - Moose needs default.  Remove specification of default = undef.
0.01  Mon Nov  2 21:51:52 2009
    - original version; created by h2xs 1.23 with options
        -Xan WebService::EastCoJp::Dictionary



lib/WebService/EastCoJp/Dictionary/GetDicItemParameter.pm
package WebService::EastCoJp::Dictionary::GetDicItemParameter;

use 5.010000;
use utf8;
use Modern::Perl;

use Moose;
use WebService::EastCoJp::Dictionary::ParameterConstraints;

has dic => (
    is      => "rw",
    isa     => "RequestParameter::Dic",
);
has item => (
    is      => "rw",
    isa     => "RequestParameter::Item",
);
has loc => (
    is      => "ro",
    isa     => "RequestParameter::Loc",
    default => q{},
);
has prof => (
    is      => "ro",
    isa     => "RequestParameter::Prof",
    default => "XHTML",
);

__PACKAGE__->meta->make_immutable;

no Moose;

our $VERSION = '0.02';

sub to_string {
    my $self = shift;
    my @attributes = qw( dic item loc prof );

    die "dic is required."
        unless $self->dic;
    die "item is required."
        unless $self->item;

    my @parameters;

    foreach my $attribute ( @attributes ) {
        my $parameter = ucfirst $attribute;

        push @parameters, join( "=", ( $parameter, $self->$attribute ) );
    }

    return join( "&", @parameters );
}

1;
__END__
=head1 NAME

WebService::EastCoJp::Dictionary::GetDicItemParameter - Make HTTP request parameter for EAST.co.jp

=head1 SYNOPSIS

  use WebService::EastCoJp::Dictionary::GetDicItemParameter;
  my $p = WebService::EastCoJp::Dictionary::GetDicItemParameter->new(
      dic  => "EJdict",
      item => "000000",
  );
  say $p->to_string;

=head1 DESCRIPTION

This module is used to make HTTP request parameter.
The parameter is for EAST.co.jp.

This module is used wish WebService::EastCoJp::Dictionary.

=head1 ATTRIBUTES

All attribute:

  - dic
  - item
  - loc
  - prof

=over

=item dic

This attribute is called Dic in EAST.co.jp.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Dic.

=item item

This attribute is called Item in EAST.co.jp.

The value of attribute is the result of the searching.
The searching request must be ealier than this attribute is set.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Item.

=item loc

This attribute is called Loc in EAST.co.jp.

This attribute offers special some behavior in some dic.
I do'nt know the value of dic.
The value might be offered to the business partner of EAST.co.jp.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Loc.

=item prof

This attribute is called Prof in EAST.co.jp.

This attribute specifies format of contents.

It is XHTML usually.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Prof.

=back

=head1 METHODS

=over

=item to_string

This method returns a string that for HTTP request.
This method will die when some attributes are undef or empty string.
The attributes are dic and item.

=back

=head1 SEE ALSO

WebService::EastCoJp::Dictionary
WebService::EastCoJp::Dictionary::ParameterConstraints
WebService::EastCoJp::Dictionary::SearchDicItemParameter

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/WebService/EastCoJp/Dictionary/ParameterConstraints.pm
package WebService::EastCoJp::Dictionary::ParameterConstraints;

use 5.010000;
use utf8;
use Modern::Perl;

use Moose::Util::TypeConstraints;

subtype "RequestParameter::Dic"
    => as "Str"
    => where {
        my $dic = $_;
        1 == grep { $dic eq $_ } qw( EJdict EdictJE wpedia );
    }
    => message { "This parameter must be EJdict or EdictJE or wpedia." }
    ;
subtype "RequestParameter::Scope"
    => as "Str"
    => where {
        my $scope = $_;
        1 == grep { $scope eq $_ } qw( HEADWORD ANYWHERE );
    }
    => message { "This parameter must be HEADWORD or ANYWHERE." }
    ;
subtype "RequestParameter::Match"
    => as "Str"
    => where {
        my $match = $_;
        1 == grep { $match eq $_ } qw( STARTWITH ENDWITH CONTAIN EXACT );
    }
    => message { "This parameter must be STARTWITH or ENDWITH or CONTAIN or EXACT." }
    ;
subtype "RequestParameter::Merge"
    => as "Str"
    => where {
        my $merge = $_;
        1 == grep { $merge eq $_ } qw( AND OR );
    }
    => message { "This parameter must be AND or OR." }
    ;
subtype "RequestParameter::Prof"
    => as "Str"
    => where { $_ eq "XHTML" }
    => message { "This parameter must be XHTML." }
    ;

subtype "RequestParameter::Item"
    => as "Str"
    => where { $_ =~ m{\A \d{4,7} \z}msx }
    => message { "This parameter must be \\d{4,7}." }
    ;
subtype "RequestParameter::Loc"
    => as "Str"
    => where { $_ eq "" }
    => message { "This parameter must be empty string." }
    ;

subtype "RequestParameter::Search"
    => as "Object"
    => where {
        eval { $_->isa( "WebService::EastCoJp::Dictionary::SearchDicItemParameter" ) };
        $@ ? 0 : 1;
    }
    => message { "This parameter must isa SearchDicItemParameter." }
    ;
subtype "RequestParameter::Get"
    => as "Object"
    => where {
        eval { $_->isa( "WebService::EastCoJp::Dictionary::GetDicItemParameter" ) };
        $@ ? 0 : 1;
    }
    => message { "This parameter must isa GetDicItemParameter." }
    ;

no Moose::Util::TypeConstraints;

our $VERSION = '0.01';

1;
__END__
=head1 NAME

WebService::EastCoJp::Dictionary::ParameterConstraints - Constraints for EAST.co.jp's parameter.

=head1 SYNOPSIS

  use Moose;
  use WebService::EastCoJp::Dictionary::ParameterConstraints;
  has dic => (
      is  => "rw",
      isa => "RequestParameter::Dic",
  );

=head1 DESCRIPTION

This module is used to for following modules.

  - WebService::EastCoJp::Dictionary
  - WebService::EastCoJp::Dictionary::SearchDicItemParameter
  - WebService::EastCoJp::Dictionary::GetDicItemParameter

=head1 CONSTRAINTS

All constraint:

  - RequestParameter::Dic
  - RequestParameter::Scope
  - RequestParameter::Match
  - RequestParameter::Merge
  - RequestParameter::Prof
  - RequestParameter::Item
  - RequestParameter::Loc
  - RequestParameter::Search
  - RequestParameter::Get

=over

=item RequestParameter::Dic

This constraint restrict to "EJdict" or "EdictJE" or "wpedia" the Dic attribute.

=item RequestParameter::Scope

This constraint restrict to "HEADWORD" or "ANYWHERE" the Scope attribute.

=item RequestParameter::Match

This constraint restrict to "STARTWITH" or "ENDWITH" or "CONTAIN" or "EXACT" the Match attribute.

=item RequestParameter::Merge

This constraint restrict to "AND" or "OR" the Merge attribute.

=item RequestParameter::Prof

This constraint restrict to "XHTML" the Prof attribute.

=item RequestParameter::Item

This constraint restrict to numeric string the Item attribute.

=item RequestParameter::Loc

This constraint restrict to empty string the Loc attribute.

=item RequestParameter::Search

This constraint restrict to SearchDicItemParameter the search attribute.

=item RequestParameter::Get

This constraint restrict to GetDicItemParameter the search attribute.

=back

=head1 SEE ALSO

WebService::EastCoJp::Dictionary
WebService::EastCoJp::Dictionary::SearchDicItemParameter
WebService::EastCoJp::Dictionary::GetDicItemParameter

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/WebService/EastCoJp/Dictionary/SearchDicItemParameter.pm
package WebService::EastCoJp::Dictionary::SearchDicItemParameter;

use 5.010000;
use utf8;
use Modern::Perl;

use Moose;
use WebService::EastCoJp::Dictionary::ParameterConstraints;

has dic => (
    is      => "rw",
    isa     => "RequestParameter::Dic",
);
has word => (
    is      => "rw",
    isa     => "Str",
);
has _word => (
    is      => "rw",
    isa     => "Str",
);
has scope => (
    is      => "rw",
    isa     => "RequestParameter::Scope",
    default => "HEADWORD",
);
has match => (
    is      => "rw",
    isa     => "RequestParameter::Match",
    default => "STARTWITH",
);
has merge => (
    is      => "rw",
    isa     => "RequestParameter::Merge",
    default => "AND",
);
has prof => (
    is      => "rw",
    isa     => "RequestParameter::Prof",
    default => "XHTML",
);
has page_size => (
    is      => "rw",
    isa     => "Int",
    default => "1",
);
has page_index => (
    is      => "rw",
    isa     => "Int",
    default => "0",
);

__PACKAGE__->meta->make_immutable;

no Moose;

use URI::Escape qw( uri_escape uri_escape_utf8 );

our $VERSION = '0.02';

sub to_string {
    my $self = shift;
    my @attributes = qw( dic word scope match merge prof page_size page_index );

    die "All parameter required."
        if grep { ( ! defined $_ ) or ( $_ eq "" ) }
           map  { $self->$_ }
           @attributes;

    my $word = $self->word;
    $self->_word(
        utf8::is_utf8( $word ) ? uri_escape_utf8( $word ) : uri_escape( $word )
    );

    my @parameters;

    MAP_FOR_REQUEST:
    foreach my $attribute ( @attributes ) {
        my $parameter = $attribute;
        my $method    = $attribute;

        # Rename parameter.
        if ( $parameter eq "word" ) {
            ( $parameter, $method ) = ( "Word", "_word" );
        }
        else {
            $parameter = join( "", map { ucfirst $_ } split "_", $parameter );
        }

        push @parameters, join( "=", ( $parameter, $self->$method ) );
    }

    return join( "&", @parameters );
}

1;
__END__
=head1 NAME

WebService::EastCoJp::Dictionary::SearchDicItemParameter - Make HTTP request parameter for EAST.co.jp

=head1 SYNOPSIS

  use WebService::EastCoJp::Dictionary::SearchDicItemParameter;
  my $p = WebService::EastCoJp::Dictionary::SearchDicItemParameter->new(
      dic  => "EJdict",
      word => "dic",
  );
  say $p->to_string;

=head1 DESCRIPTION

This module is used to make HTTP request parameter.
The parameter is for EAST.co.jp.

This module is used with WebService::EastCoJp::Dictionary.

=head1 ATTRIBUTES

All attribute:

  - dic
  - word
  - scope
  - match
  - merge
  - prof
  - page_size
  - page_index

=ober

=item dic

This attribute is called Dic in EAST.co.jp.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Dic.

=item word

This attribute is called Word in EAST.co.jp.

This attribute specifies the word that will be translated.

This will encode when call to_string method.
The encode is return value of URI::Escape::uri_encode.

=item scope

This attribute is called Scope in EAST.co.jp.

This attribute specifies a space.
The space is subject or anywhere.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Scope.

=item match

This attribute is called Match in EAST.co.jp.

This attribute specifies position to which position of the word
the character is corresponding.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Match.

=item merge

This attribute is called Merge in EAST.co.jp.

This attribute specifies how the query is joined.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Merge.

=item prof

This attribute is called Prof in EAST.co.jp.

This attribute specifies format of contents.

It is XHTML usually.

This attribute has constraint of WebService::EastCoJp::Dictionary::ParameterConstraints's
RequestParameter::Prof.

=item page_size

This attribute is called PageSize in EAST.co.jp.

This attribute specifies the size of result set.

=item page_index

This attribute is called PageIndex in EAST.co.jp.

This attribute specifies the start index
that is returned value of search request.
It is zero usually.

=back

=head1 METHODS

=over

=item to_string

This method returns a string that for HTTP request.
This method will die when some attributes are undef or empty string.
The attributes are dic and word.

=back

=head1 SEE ALSO

WebService::EastCoJp::Dictionary
WebService::EastCoJp::Dictionary::ParameterConstraints
WebService::EastCoJp::Dictionary::GetDicItemParameter

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/WebService/EastCoJp/Dictionary.pm
package WebService::EastCoJp::Dictionary;

use 5.010000;
use utf8;
use Modern::Perl;
use Moose;
use WebService::EastCoJp::Dictionary::ParameterConstraints;
use WebService::EastCoJp::Dictionary::SearchDicItemParameter;
use WebService::EastCoJp::Dictionary::GetDicItemParameter;

has parameter => (
    is      => "rw",
    isa     => "RequestParameter::Search",
    default => sub { WebService::EastCoJp::Dictionary::SearchDicItemParameter->new( ) },
);

__PACKAGE__->meta->make_immutable;

no Moose;

use Carp;
use Readonly;
use LWP::UserAgent;
use XML::XPath;

our $VERSION = '0.02';

Readonly my %URL => (
    search_item => "http://btonic.est.co.jp/NetDic/NetDicV09.asmx/SearchDicItemLite",
    get_item    => "http://btonic.est.co.jp/NetDic/NetDicV09.asmx/GetDicItemLite",
);

sub can_reach_to_the_server {
    my $self    = shift;
    my $ua      = LWP::UserAgent->new( );
    my $message = "500 Internal Server Error";

    CAN_NOT_REACH:
    foreach my $url ( values %URL ) {
        my $res = $ua->get( $url );

        return
            unless lc $res->status_line eq lc $message;
    }

    return 1;
}

sub _search_dic_item_lite {
    my $self = shift;

    my $p   = $self->parameter;
    my $url = join( "?", ( $URL{search_item}, $p->to_string ) );
    my $ua  = LWP::UserAgent->new( );

    my $res = $ua->get( $url );
    die "Could not search item.\n[", $res->status_line, "]"
        if $res->is_error;

    my $content = $res->decoded_content;

    # The content has CRLF, but I want LF.
    $content =~ tr{\x0d}{}d;
#say "### Searched XML ", "#" x 62, "\n", $content, "\n", "#" x 80;

    # Die if no item.
    my $xpath = XML::XPath->new( xml => $content );
    my( $hit_node ) = $xpath->find( "//TotalHitCount" )->get_nodelist;
    die $self->parameter->word, " is not registered in ", $self->parameter->dic, ".\n"
        if $hit_node->string_value eq "0";

    return $content;
}

sub _get_dic_item_lite {
    my $self = shift;

    die "An argument required."
        if @_ != 1;

    my $p = shift;
    my $p_isa_get_parameter
        = eval { $p->isa( "WebService::EastCoJp::Dictionary::GetDicItemParameter" ) };

    die "An argument must isa GetDicItemParameter."
        unless $p_isa_get_parameter;

    my $url = join( "?", ( $URL{get_item}, $p->to_string ) );
    my $ua  = LWP::UserAgent->new( );

    my $res = $ua->get( $url );
    die "Could not get item.\n[", $res->status_line, "]"
        if $res->is_error;

    my $content = $res->decoded_content;

    # The content has CRLF, but I want LF.
    $content =~ tr{\x0d}{}d;

    return $content;
}

sub _die_if_xml_has_error {
    my $self = shift;
    my $xml  = shift;

    my $xpath    = XML::XPath->new( xml => $xml );
    my( $error ) = $xpath->find( "//ErrorMessage" )->get_nodelist;
    my $message  = $error->toString;
#say "### ", $error->toString;

    die "Received XML has error[$message]."
        if $message ne "<ErrorMessage />";

    return;
}

sub _search_and_get_dic_item {
    my $self = shift;

    croak "Argument does not has ref."
        if @_ == 1 and ref $_[0] eq ref +{ };
    croak "The number of argument is odd.  Argument can not be treated hash."
        if @_ % 2;

    my %parameter = @_ != 1 ? @_ : %{ $_[0] };
    my $p         = $self->parameter;
    # TODO: Change this copy to shallow copy.
    # This will change default parameter without user's Agreement.
    # It does not care now.  Because this changes arguments only
    # dic and word.  Be carefull when this changes other arguemtns.

    SET_PARAMETER:
    foreach my $key ( sort keys %parameter ) {
        $p->$key( $parameter{ $key } );
    }

    my $xml   = $self->_search_dic_item_lite;
    my $xpath = XML::XPath->new( xml => $xml );
#say "### xml ", "#" x 72;
#say $xml;
#say "#" x 80;

    # Did fail?
    $self->_die_if_xml_has_error( $xml );

    # Set with narrow scope.
    my %translation;
    {
        my @items  = map { $_->string_value }
                     $xpath->find( "//DicItemTitle/ItemID" )->get_nodelist;
        my @titles = map { $_->string_value }
                     $xpath->find( "//DicItemTitle/Title/span" )->get_nodelist;
        die "Could not parse XML."
            unless @items == @titles;

        %translation = map { ( $titles[$_], $items[$_] ) }
                       ( 0 .. $#items );
    }

    my $get_p = WebService::EastCoJp::Dictionary::GetDicItemParameter->new( );
    $get_p->dic( $p->dic );

    my @xmls;

    GET_XMLS:
    foreach my $title ( sort keys %translation ) {
        my $item = $translation{ $title };
        $get_p->item( $item );

        my $xml = $self->_get_dic_item_lite( $get_p );
#say "### xml ", "#" x 72;
#say $xml;
#say "#" x 80;

        # Did fail?
        $self->_die_if_xml_has_error( $xml );

        if ( ! wantarray ) {
            return $xml;
        }
        else {
            push @xmls, $xml;
        }
    }

    return @xmls;
}

sub enja {
    my $self  = shift;
    my @words = @_;
    croak "Argument required."
        unless @words;

    $self->parameter->dic( "EJdict" );
    $self->parameter->word( join( " ", @words ) );
    # TODO: This changes global parameter, but it does not notice to user.
    # Consider this when another parameter will be changed.

    my @xmls = wantarray ? $self->_search_and_get_dic_item
                         : ( scalar $self->_search_and_get_dic_item )
                         ;
#print map { "#" x 80 . "\n" . $_ . "\n" . "#" x 80 . "\n" } @xmls;

    my %translation;

    PARSE_TRANSLATED_XML:
    foreach my $xml ( @xmls ) {
        my $xpath = XML::XPath->new( xml => $xml );

        my( $head_node ) = $xpath->find( "//Head/div/span" )->get_nodelist;
        my( $body_node ) = $xpath->find( "//Body/div/div" )->get_nodelist;

        $translation{ $head_node->string_value }
            = $body_node->string_value;
    }

    if ( ! wantarray ) {
        my( $content ) = map { join( "\t", ( $_, $translation{ $_ } ) ) }
                         sort keys %translation
                         ;
        return $content;
    }
    else {
        return map { join( "\t", ( $_, $translation{ $_ } ) ) }
               sort keys %translation
               ;
    }
}

sub jaen {
    my $self  = shift;
    my @words = @_;
    croak "Argument required."
        unless @words;

    $self->parameter->dic( "EdictJE" );
    $self->parameter->word( join( " ", @words ) );
    # TODO: This changes global parameter, but it does not notice to user.
    # Consider this when another parameter will be changed.

    my @xmls = wantarray ? $self->_search_and_get_dic_item
                         : ( scalar $self->_search_and_get_dic_item )
                         ;
#print map { "#" x 80 . "\n" . $_ . "\n" . "#" x 80 . "\n" } @xmls;

    my %translation;

    PARSE_TRANSLATED_XML:
    foreach my $xml ( @xmls ) {
        my $xpath = XML::XPath->new( xml => $xml );

        my( $head_node ) = $xpath->find( "//Head/div/span" )->get_nodelist;
        my @body_nodes   = $xpath->find( "//Body/div/div/div" )->get_nodelist;

        my $heading = $head_node->string_value;
        $heading = join( "", split "\n", $heading );
        $heading = join( "ー", split "[\\s]+", $heading );

        my @bodies = grep { $_ !~ m{\A \s* \z}msx }
                     map  { $_->string_value }
                     @body_nodes
                     ;
        my $body = join( "\t", @bodies );

        $translation{ $heading } = $body;
    }

    if ( ! wantarray ) {
        my( $content ) = map { join( "\t", ( $_, $translation{ $_ } ) ) }
                         sort keys %translation
                         ;
        return $content;
    }
    else {
        return map { join( "\t", ( $_, $translation{ $_ } ) ) }
               sort keys %translation
               ;
    }
}

sub jaja {
    my $self  = shift;
    my @words = @_;
    croak "Argument required."
        unless @words;

    $self->parameter->dic( "wpedia" );
    $self->parameter->word( join( " ", @words ) );
    # TODO: This changes global parameter, but it does not notice to user.
    # Consider this when another parameter will be changed.

    my @xmls = wantarray ? $self->_search_and_get_dic_item
                         : ( scalar $self->_search_and_get_dic_item )
                         ;
#print map { "#" x 80 . "\n" . $_ . "\n" . "#" x 80 . "\n" } @xmls;

    my %translation;

    PARSE_TRANSLATED_XML:
    foreach my $xml ( @xmls ) {
        my $xpath = XML::XPath->new( xml => $xml );

        my( $head_node ) = $xpath->find( "//Head/div/span" )->get_nodelist;
        my( $body_node ) = $xpath->find( "//Body/div/div" )->get_nodelist;

        $translation{ $head_node->string_value }
            = $body_node->string_value;
    }

    if ( ! wantarray ) {
        my( $content ) = map { join( "\t", ( $_, $translation{ $_ } ) ) }
                         sort keys %translation
                         ;
        return $content;
    }
    else {
        return map { join( "\t", ( $_, $translation{ $_ } ) ) }
               sort keys %translation
               ;
    }
}

1;
__END__
=encoding utf-8

=head1 NAME

WebService::EastCoJp::Dictionary - An interface for REST service of EAST.co.jp

=head1 SYNOPSIS

  use WebService::EastCoJp::Dictionary;

  die "Could not reach to the server."
      unless WebService::EastCoJp::Dictionary::can_reach_to_the_server( );

  my $d = WebService::EastCoJp::Dictionary->new( );
  $d->parameter->match( "EXACT" );

  say $d->enja( "English" );
  say $d->jaen( "日本語" );

  my( $title, @contents ) = split "\t", $d->enja( "Japanese" );

  $d->parameter->match( "STARTWITH" );
  $d->parameter->page_size( 3 );

  SAY_TRANSLATION:
  foreach my $translation ( $d->enja( "loc" ) ) {
      my( $title, @contents ) = split "\t", $translation;

      my $head_message = "### $title ";
      my $length = length( $head_message )
                   + do { require bytes; bytes::length( $head_message ) };
      $length /= 2;
      say $head_message, "#" x ( 80 - $length );
      print map { "[ $_ ]\n" } @contents;
  }

=head1 DESCRIPTION

This module is an interface for REST service of EAST.co.jp.
The server offers the dictionary retrieval.

The server has URL of "http://www.btonic.com/ws/index.html".

The server has some services, which are:

  - translate English into Japanese
  - translate Japanese into English

=head1 ATTRIBUTES

This class has an attribute.
The attribute isa WebService::EastCoJp::Dictionary::SearchDicItemParameter.

=head1 METHODS

Methods are:

  - can_reach_to_the_server: test it
  - enja: for translate English into Japanese
  - jaen: for translate Japanese into English

Following methods are for developer:

  - _search_dic_item_lite
  - _get_dic_item_lite

=over

=item can_reach_to_the_server

This method tests that this environment can reach to
the server of EAST.co.jp.

=item enja

This method returns the translation of argument.
This method will die when it fails in translation.

The returned value is Japanese.

=item jaen

This method returns the translation of argument.
This method will die when fail translation.

The returned value is English.

=item jaja

This method has been reserved for the future.

This will describe a argument.

=back

=head1 SEE ALSO

WebService::EastCoJp::Dictionary::SearchDicItemParameter
WebService::EastCoJp::Dictionary::GetDicItemParameter
WebService::EastCoJp::Dictionary::ParameterConstraints

=head1 AUTHOR

Kuniyoshi Kouji, E<lt>kuniyoshi@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Kuniyoshi Kouji

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut

